### ISSUE TITLE
Add gpt_oss architecture support (gpt-oss-20b) incl. MXFP4 + official safetensors loading

### ISSUE THREAD

**Issue by pekkah** [pekkah/SharpInference#182]
## Summary

Add support for OpenAI's `gpt_oss` architecture so SharpInference can run **gpt-oss-20b** (and, with expert offload, gpt-oss-120b). gpt-oss-20b is a 21B-total / 3.6B-active MoE under Apache 2.0 — a strong reasoning/agentic model that fits the **12 GB VRAM + 64 GB RAM** target via our existing MoE expert-streaming path. It's an OpenAI-weights lineage we have no coverage for today, distinct in strengths from the Qwen/Gemma/Llama models we already run.

This issue is intentionally scoped to cover **both**:
- **Workstream A** — the `gpt_oss` transformer architecture + the **MXFP4** quant format it requires.
- **Workstream B** — a first-class **safetensors LLM loader** so we can consume the **official release** directly, instead of depending on third-party GGUF conversions.

> Config values below are **confirmed** against the official HF `transformers` `GptOssConfig` and OpenAI's reference implementation ([openai/gpt-oss `model.py`](https://github.com/openai/gpt-oss/blob/main/gpt_oss/torch/model.py)). No open verify-flags remain.

---

## Model spec (gpt-oss-20b)

| Field | Value |
|---|---|
| Total / active params | 21B / 3.6B |
| Layers | 24 |
| hidden_size | 2880 |
| Attention heads / head_dim | 64 / 64 |
| KV heads (GQA) | 8 |
| Experts / top-k | 32 / 4 (softmax router) |
| Expert FFN intermediate | 2880 |
| Activation | **clamped gated SwiGLU** (see below), `alpha=1.702`, `swiglu_limit=7.0`, **with expert bias terms** |
| Attention pattern | **alternating** sliding-window / full attention per layer |
| Sliding window | 128 tokens |
| Attention sinks | **yes** — learned per-head sink logit in the softmax denominator |
| RoPE | NEOX, `rope_theta=150000`, **YaRN** (`factor=32`, `beta_fast=32`, `beta_slow=1`, `original_max_position_embeddings=4096`) |
| Context | 131,072 |
| attention_bias | true (biased Q/K/V/O projections) |
| Vocab | 201,088 (o200k_harmony) |
| Weights | **MXFP4** expert tensors (FP4, block 32, e8m0 shared exponent) |

### Exact activation (don't get this wrong)
gpt-oss does **not** use plain SwiGLU. Per the reference impl:
```
gate = clamp(gate, max=+7.0)          # upper-bound only (asymmetric)
up   = clamp(up,   min=-7.0, max=+7.0) # symmetric clamp
out  = (up + 1.0) * gate * sigmoid(1.702 * gate)
```
Note the **`+1.0` on the linear branch** (linear defaults to identity) and the **asymmetric clamp** on `gate`. The clamps exist to keep activations stable in MXFP4. Experts also carry **bias terms** on gate/up/down.

---

## Sizing / format reality (why MXFP4 is mandatory)

The model is **~12 GB in every GGUF flavor** — there is no smaller "Q4" target:

| Variant | Size | Notes |
|---|---|---|
| native MXFP4 (safetensors, official) | 12.1 GB | OpenAI release format |
| community "Q4_K_M" GGUF | 11.9 GB | **experts still MXFP4**; only attn/embed tensors differ |
| Q5_K_M / Q6_K / Q8_0 | ~11.5–13 GB | all cluster ~12 GB |
| F16/BF16 upcast | ~40 GB | not useful |

The MoE expert FFN tensors dominate size and ship in MXFP4; quantizers keep them MXFP4 because requantizing gains nothing and hurts quality. **Consequence: MXFP4 dequant is on the critical path — even a file named `…Q4_K_M.gguf` contains MXFP4 expert tensors, so we cannot load any practical gpt-oss GGUF (or the official safetensors) without an MXFP4 kernel.**

At ~12 GB the model slightly overruns 12 GB VRAM once KV cache is added → route experts to the 64 GB RAM tier via `ExpertSlotManager` (our existing offload path).

### Note on "official vs GGUF"
The official `openai/gpt-oss-20b` release is **safetensors + MXFP4** (no official GGUF). Community GGUFs (ggml-org/unsloth/bartowski) **preserve the MXFP4 expert tensors bit-for-bit** — they are a repackaging of the official expert weights, not a requant of them. So GGUF and official safetensors yield the same expert weights; the MXFP4 kernel work is identical either way. Workstream B (safetensors) is justified by **day-one model support + independence from third-party converters**, not by gpt-oss fidelity — gpt-oss is simply its first validation consumer.

---

## What we already have (reusable)

**Core LLM stack (GGUF):**
- ✅ NEOX RoPE — `gpt-oss` is already in the NEOX switch (`ModelGraph.cs:311`).
- ✅ MoE detection, routing, top-k, softmax gating, shared-expert path — `ModelGraph.cs:256-257`, `ForwardPass.cs:1705-1782` (`MoeFfn`).
- ✅ Per-layer sliding-window attention + mask — `ForwardPass.cs:1539` (Gemma 4 path); per-layer config plumbing exists.
- ✅ GQA — `ForwardPass.cs:1548-1599`.
- ✅ Arch-prefixed metadata helpers — `ModelGraph.cs:533-586`.

**Safetensors (Diffusion only, needs promotion + extension):**
- ✅ `IWeightLoader` abstraction + `SafetensorsLoader` (single + sharded + `index.json`) — `src/SharpInference.Diffusion/`.
- ⚠️ Lives in `SharpInference.Diffusion`, not Core.
- ⚠️ `ReadF32`-centric; `TryGetRaw` returns `false` (`SafetensorsLoader.cs:238`) — no mmap/raw-quant path for the LLM hot path.
- ⚠️ Handles F32/F16/BF16/F8 only (`SafetensorsLoader.cs:128-150`) — **no MXFP4**.
- ⚠️ No HF→internal tensor-name mapping, no `config.json` hyperparam path, no standalone `tokenizer.json` loading.

---

## Workstream A — `gpt_oss` architecture + MXFP4

1. **`gpt_oss` dispatch + hyperparams** (`ModelGraph.cs:254+`): read `gpt-oss.*` keys; build alternating sliding/full per-layer pattern (`layer_types`); wire YaRN rope scaling (`factor=32`, `beta_fast=32`, `beta_slow=1`, `original_max_position_embeddings=4096`, `rope_theta=150000`); head/vocab config.
2. **MXFP4 DType + dequant** *(critical path)*:
   - Add `MXFP4` to `DType` enum + block/byte info (`Tensor.cs:43-81`; block=32, e8m0 shared exponent).
   - CPU `DequantMXFP4` + `MatVecMXFP4` (`Dequantize.cs`, `SimdKernels.cs:124-148` dispatch).
   - GPU dequant kernels (CUDA/Vulkan) for offloaded-expert and GPU forward paths.
3. **Attention sinks** (`ForwardPass.cs:1526-1600`): append a per-head learned sink logit to the softmax denominator (`SoftmaxInPlace` at `:1571`); sink contributes no value. New per-head weight load.
4. **Clamped gated SwiGLU for experts** (`ForwardPass.cs:1748,1775`): experts are currently hardcoded to `SiLuMul`; add the exact gpt-oss variant `out = (clamp(up,-7,7) + 1) * clamp(gate,max=7) * sigmoid(1.702*gate)` **plus expert bias terms** (gate/up/down), which `MoeFfn`'s `FusedMatVec` path does not currently add. Note the asymmetric clamp on `gate` and the `+1` on `up`.
5. **Attention/QKV bias**: gpt-oss uses biased projections (`attention_bias=true`) — confirm bias load path covers attn + router.

## Workstream B — safetensors LLM loader (consume the official release)

1. **Promote loader to Core** (or a shared package) so `GgufModel` and a new `SafetensorsModel` sit behind a common interface feeding `ModelHyperparams` + the forward pass.
2. **Raw / mmap path + MXFP4** in the safetensors reader so large quant tensors feed `MatVec` directly (do **not** expand ~12 GB MXFP4 → ~80 GB f32). Reuse the Workstream A MXFP4 kernel. Handle HF's MXFP4 `blocks` + `scales` paired-tensor layout.
3. **HF tensor-name mapping** (`model.layers.N.self_attn.q_proj.weight` → internal `blk.N.attn_q.weight` conventions).
4. **`config.json` hyperparam source** feeding `ModelHyperparams` (parallel to GGUF metadata keys).
5. **`tokenizer.json` loading** (Microsoft.ML.Tokenizers) incl. special tokens / chat template — for gpt-oss this is the harmony/o200k tokenizer.

## Out of scope (explicitly)
- PyTorch `.bin`/`.pth` (pickle): security (arbitrary code exec) + reflection/dynamic — incompatible with our NativeAOT/trim constraints.
- ONNX: separate graph/runtime paradigm; no alignment with our hand-written kernels.

---

## Phasing

- **Phase 1 — arch correctness (CPU):** Workstream A items 1,3,4,5 + **CPU** MXFP4 dequant. Validate against an MXFP4-preserving GGUF (reuses the whole existing Core loader/tokenizer/metadata stack → fastest path to a working model).
- **Phase 2 — official safetensors:** Workstream B — load `openai/gpt-oss-20b` safetensors directly; validate numerics against Phase 1.
- **Phase 3 — performance:** GPU MXFP4 kernels + expert offload via `ExpertSlotManager` to hit the 12 GB / 64 GB target at usable tok/s; optional MXFP4-direct matvec (skip full dequant).

---

## Acceptance criteria

- [ ] gpt-oss-20b GGUF loads; `hp` reports MoE, alternating SWA pattern, attention sinks, YaRN.
- [ ] CPU forward pass produces finite, coherent greedy decode (new test in `Tests.ForwardPass`, mirroring `Gemma4CpuForwardPassTests`).
- [ ] Greedy output matches llama.cpp reference within argmax stability on a fixed prompt.
- [ ] Official `openai/gpt-oss-20b` **safetensors** loads via the Core safetensors path and matches GGUF numerics.
- [ ] Runs within 12 GB VRAM + 64 GB RAM via expert offload at usable tok/s.
- [ ] `dotnet build` clean (TreatWarningsAsErrors), NativeAOT-compatible (no reflection/dynamic added).

---

## References

- [openai/gpt-oss-20b](https://hf.co/openai/gpt-oss-20b) · [config.json](https://huggingface.co/openai/gpt-oss-20b/blob/main/config.json) · [reference impl `model.py`](https://github.com/openai/gpt-oss/blob/main/gpt_oss/torch/model.py) · [HF `transformers` GptOss docs](https://huggingface.co/docs/transformers/en/model_doc/gpt_oss)
- [gpt-oss tech report (arXiv 2508.10925)](https://arxiv.org/abs/2508.10925) · [gpt-oss from the ground up (Cameron Wolfe)](https://cameronrwolfe.substack.com/p/gpt-oss) · [What's different about GPT-OSS SwiGLU (ceramic.ai)](https://www.ceramic.ai/blog/gpt-oss-swiglu)
- [bartowski/openai_gpt-oss-20b-GGUF](https://hf.co/bartowski/openai_gpt-oss-20b-GGUF) (size table) · [MXFP4 experimental](https://hf.co/bartowski/openai_gpt-oss-20b-GGUF-MXFP4-Experimental)
- [Unsloth gpt-oss run guide](https://unsloth.ai/docs/models/gpt-oss-how-to-run-and-fine-tune)
